src/library/scala/collection/View.scala

```diff
@@ -163,6 +163,33 @@ object View extends IterableFactory[View] {
     def iterator = partition.underlying.iterator.filter(x => partition.p(x) == cond)
   }
 
+  /** A view that splits an underlying collection into two views */
+  @SerialVersionUID(3L)
+  class MapSplit[A, A1, A2](val underlying: SomeIterableOps[A], val f: A => Either[A1, A2]) {
+
+    /** The view consisting of all elements of the underlying collection
+      *  that map to `Left`.
+      */
+    val left = new Split[A,A1,A2,A1](this, new PartialFunction[A, A1] {
+      override def isDefinedAt(a: A): Boolean = f(a).isLeft
+      override def apply(a: A): A1 = f(a).left.get
+    })
```

Comment:
What do you think of implementing our own `Iterator` instead of using `collect`?

Reply:
Not sure I really get it, do you mean something like:
```
class MapSplit[A, A1, A2](val underlying: SomeIterableOps[A], val f: A => Either[A1, A2]) {
  def iterator = underlying.iterator.mapSplit(f)
 }
```

Where `mapSplit` is implemented in `Iterator`?

Reply:
We can not implement `mapSplit` in `Iterator`, but the beginning of your snippet is indeed what I have in mind. Then you can implement an anonymous `Iterator`. I’m not sure this is a best way to go, but the overhead of using these partial functions doesn’t sound very appealing.

Reply:
Hey @valydia, sorry my comment was not detailed enough.

Here is a sketch of what I have in mind:

~~~ scala
  class MapSplit[A, A1, A2](val underlying: SomeIterableOps[A], val f: A => Either[A1, A2]) {

    val left: Iterator[A1] = new AbstractIterator[A1] {
      private val self = underlying.iterator
      private var hd: A1 = _
      private var hdDefined: Boolean = false
      def hasNext = hdDefined || {
        def findNext(): Boolean =
          if (self.hasNext) {
            f(self.next()) match {
              case Left(a1) => hd = a1; hdDefined = true; true
              case Right(_) => findNext()
            }
          } else false
        findNext()
      }
      def next() =
        if (hasNext) {
          hdDefined = false
          hd
        } else Iterator.empty.next()
    }

    val right: Iterator[A2] = new AbstractIterator[A2] {
      // Similar implementation as for `left`
    }
}
~~~

(I haven’t tested this code)

Reply:
Thanks for clarifying @julienrf! The trick was to duplicate the code, for `Left` and `Right`! One question though, shouldn't `Left` and `Right` be instance of  `AbstractView`?

Reply:
Thanks @julienrf  for clarifying, the trick was indeed to duplicate the code for `left` and `right`! One question though, Don't `left` and `right` need to be `AbstractView` rather than `Iterator`?